### PR TITLE
Pipes - Sara Frandsen - Media Ranker (late)

### PR DESCRIPTION
**Pull request by frankienakama**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In self.top_ten(category), it orders each work category by votes in descending order, and returns the first ten in the list. This gives us the top ten works for each category.
Describe how you approached testing that model method. What edge cases did you come up with? | TBH I did not do any testing.
Describe an edge case test you wrote for a controller |
What are `session` and `flash`? What is the difference between them? | Sessions act like hashes that sort of stick around without needing to exist in the database. Flash objects are the same, but disappear after a single request.
Describe a controller filter you wrote. |
What was one thing that you gained more clarity on through this assignment? | Using the model the write methods in! USING GIT AND FIXING THE THINGS I BREAK
What is the Heroku URL of your deployed application | https://sara-frandsen-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'works#main'
+
+  resources :users, only: [:index, :show]
+  resources :sessions
```
Comment (droberts-sea) [inline config/routes.rb:7]: I'm not sure if you need the full `resources` for `:sessions` here, especially since you've got login/logout listed explicitly below.

File: app/views/works/_top-ten.html.erb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<%# partial view for top 10 lists %>
+<article>
```
Comment (droberts-sea) [inline app/views/works/_top-ten.html.erb:2]: Great use of a partial to DRY this up!

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class User < ApplicationRecord
+  has_many :votes
+
+  def upvote(work_id)
+    work = Work.find(work_id)
+    if Vote.where(work: work, user: self).count > 0
+      return false, 'Cannot upvote more than once!'
```
Comment (droberts-sea) [inline app/models/user.rb:7]: I like this as a model method! Good work!

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      redirect_to work_path(@work.id)
+    else
+      render new_work_path
+    end
+  end
+
+  def show
+    # individual work
+    @work = Work.find(params[:id])
+  end
```
Comment (droberts-sea) [inline app/controllers/works_controller.rb:39]: This would be a great bit of code to DRY up using a controller filter, especially if you extend the functionality to return `404 not_found` if that model instance doesn't exist.

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes - some extra routes
Semantic HTML | yes
Errors are reported to the user | yes
Business logic lives in the models | yes - good work
Models are thoroughly tested, including relations, validations and any custom logic | no
Controllers are thoroughly tested | no
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
Media pages contain lists of voting users | yes
**Wave 3 - Styling** |
Foundation is used appropriately | some - seems like there's still a lot of ground to explore here
Look and feel is similar to the original | some
**Overall** | Good job overall. This is a big, difficult project, and while some of the pieces weren't quite there (particularly around styling and testing), you were able to get the core functionality working. This is no mean feat. Make sure you take the opportunity to practice these on the bEtsy project, and keep up the hard work!
